src/analyzer/analyzer.h

```diff
@@ -15,12 +14,71 @@
 
 class Analyzer {
   public:
+    virtual ~Analyzer() = default;
+
+    // This method is supposed to:
+    //  1. Check if the track needs to be analyzed, otherwise return false.
+    //  2. Perform the initialization and return true on success.
+    //  3. If the initialization failed log the internal error and return false.
     virtual bool initialize(TrackPointer tio, int sampleRate, int totalSamples) = 0;
-    virtual bool isDisabledOrLoadStoredSuccess(TrackPointer tio) const = 0;
-    virtual void process(const CSAMPLE* pIn, const int iLen) = 0;
-    virtual void cleanup(TrackPointer tio) = 0;
+
+    /////////////////////////////////////////////////////////////////////////
+    // All following methods will only be invoked after initialize()
+    // returned true!
+    /////////////////////////////////////////////////////////////////////////
+
+    // Analyze the next chunk of audio samples and return true if successful.
+    // If processing fails the analysis can be aborted early by returning
+    // false. After aborting the analysis only cleanup() will be invoked,
+    // but not finalize()!
+    virtual bool process(const CSAMPLE* pIn, const int iLen) = 0;
+
+    // Update the track object with the analysis results after
+    // processing finished successfully, i.e. all available audio
+    // samples have been processed.
     virtual void finalize(TrackPointer tio) = 0;
-    virtual ~Analyzer() = default;
+
+    // Discard any temporary results or free allocated memory.
+    virtual void cleanup() = 0;
 };
 
-#endif
+typedef std::unique_ptr<Analyzer> AnalyzerPtr;
+
+class AnalyzerState {
+  public:
+    explicit AnalyzerState(AnalyzerPtr analyzer)
+            : m_analyzer(std::move(analyzer)),
+              m_processing(false) {
+        DEBUG_ASSERT(m_analyzer);
+    }
+    AnalyzerState(const AnalyzerState&) = delete;
+    AnalyzerState(AnalyzerState&&) = default;
+
+    bool initialize(TrackPointer tio, int sampleRate, int totalSamples) {
+        DEBUG_ASSERT(!m_processing);
+        m_processing = m_analyzer->initialize(tio, sampleRate, totalSamples);
+        return m_processing;
+    }
+
+    void process(const CSAMPLE* pIn, const int iLen) {
+        if (m_processing) {
+            m_processing = m_analyzer->process(pIn, iLen);
+        }
+    }
+
+    void finalize(TrackPointer tio) {
+        if (m_processing) {
+            m_analyzer->finalize(tio);
```

Comment:
if process() fails in one call finalize is skipped. Can this be an issue?

Reply:
finalize() is only supposed to be invoked if process() finished successfully. It doesn't make sense to store incomplete or invalid results. The documentation of finalize() explicitly states this:
"Update the track object with the analysis results after processing **finished successfully**..."

Reply:
ÒK, I am unsure if the is obvious everywhere. finalize() reads for me as if the analyzer PlugIn itself is responsible for this. I think this has changed now due to the external AnalyzerState wrapper. 
Does it make sense to rename this like storeResults().